### PR TITLE
Added rate limits in upgrade

### PR DESCRIPTION
**Pull request by nicolaslara**
Closes: #4303 

## What is the purpose of the change

This adds the rate limits in the v15 upgrade for when https://www.mintscan.io/osmosis/proposals/427 passes.

## Brief Changelog

 * Add rate limits in v15

## Testing and Verifying

 - [x] The upgrade has tests.
 - [x] The upgrade has E2E tests
 - [ ] This should also be tested on a testnet with state export. 

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

**Comment by ValarDragon** [comment]
RIP goimports error. Do you have your editor set to format on save?

**Comment by nicolaslara** [comment]
> RIP goimports error. Do you have your editor set to format on save?

I do, but it doesn't work when I fix conflicts and some other weird situations like that.


@ValarDragon  but!! this is not ready. The E2E tests are passing because I deactivated them after failing to get them to work (rip).

I asked Nicco to test the upgrade with a state export to make sure the issue is with E2E and not with the upgrade

**Comment by nicolaslara** [comment]
That change makes a lot of sense! I thought that was just initialization. Thanks for the find!

**Comment by nicolaslara** [comment]
hmm, this is still failing when there contracts exist on v14. I'll try to test this again tomorrow on a real node and see what happens

**Comment by p0mvn** [comment]
Latest updates:
- wired ibc-rate-limit as SDK module to have init and export genesis logic
- changed querier to be code-gen compatible and fixed CLI
-  fixed bug with not getting params correctly due to not having a pointer receiver on `ParamSetPairs`: https://github.com/osmosis-labs/osmosis/blob/af89301de5270d9bb3f4b6e39c12d407ca331270/x/ibc-rate-limit/types/params.go#L44
- fixed bug with overriding contract parameter due to `ibc-rate-limit` consensus version being set to 1, leading to overwrite during the upgrade. Fixed with: https://github.com/osmosis-labs/osmosis/blob/af89301de5270d9bb3f4b6e39c12d407ca331270/app/upgrades/v15/upgrades.go#L56

We manually checked the presence of the contract parameter and rate limits post-upgrade. However, we should add a test in `e2e_test.go` doing that for us. I think this is the only remaining TODO here

File: tests/e2e/configurer/chain/chain.go [hunk 1]
```diff
@@ -43,6 +53,8 @@ const (
 	waitUntilRepeatPauseTime = 2 * time.Second
 	// waitUntilrepeatMax is the maximum number of times to repeat the wait until condition.
 	waitUntilrepeatMax = 60
+
+	proposalStatusPassed = "PROPOSAL_STATUS_PASSED"
```
Comment (nicolaslara) [inline tests/e2e/configurer/chain/chain.go:57]: This seems unrelated to this PR. Do we want it here?

Comment (p0mvn) [reply tests/e2e/configurer/chain/chain.go:56]: This was used to ensure that the param change proposal passed before we tried to get the param itself. This could help to be more confident that the proposal not passing was the reason for the parameter not updating. However, that check seems to be removed now.

I'm cool with not blocking on this and just deleting `prposalStatusPassed`

Comment (p0mvn) [reply tests/e2e/configurer/chain/chain.go:56]: Ahh, actually it's still being used for checking that the proposal passed here: https://github.com/osmosis-labs/osmosis/blob/376fccde469c3c4168127e48ee9beda44cee7647/tests/e2e/configurer/chain/chain.go#L285

So I think we can keep this

File: tests/e2e/e2e_test.go [hunk 2]
```diff
@@ -445,23 +444,29 @@ func (s *IntegrationTestSuite) TestSuperfluidVoting() {
...
+	s.Require().NoError(err)
+
+	// Need to json unparshal the params because they are stored including quotes
+	paramA := nodeA.QueryParams(ibcratelimittypes.ModuleName, string(ibcratelimittypes.KeyContractAddress))
+	json.Unmarshal([]byte(paramA), &paramA)
+	paramB := nodeB.QueryParams(ibcratelimittypes.ModuleName, string(ibcratelimittypes.KeyContractAddress))
+	json.Unmarshal([]byte(paramB), &paramB)
+
+	// When upgrading to v15, we want to make sure that the rate limits have been set.
+	_, err = nodeA.QueryWasmSmartArray(paramA, `{"get_quotas": {"channel_id": "any", "denom": "ibc/E6931F78057F7CC5DA0FD6CEF82FF39373A6E0452BF1FD76910B93292CF356C1"}}`)
```
Comment (p0mvn) [inline tests/e2e/e2e_test.go:468]: Do we want to check that the values (or at least some of them) are set correctly?

File: tests/e2e/configurer/chain/chain.go (hunk 1 above)
Comment (nicolaslara) [reply tests/e2e/configurer/chain/chain.go:56]: great!

File: tests/e2e/e2e_test.go (hunk 2 above)
Comment (nicolaslara) [reply tests/e2e/e2e_test.go:468]: good call! I'll check the last one (like on the test). If any of the other ones fail, it should panic before even executing that.

Comment (nicolaslara) [reply tests/e2e/e2e_test.go]: actually, that query already fetches the latest value. if the upgrade failed that should be empty (`ibc/E6931F78057F7CC5DA0FD6CEF82FF39373A6E0452BF1FD76910B93292CF356C1` doesn't exist).  Added a length check

File: tests/e2e/e2e_test.go [hunk 3]
```diff
@@ -445,23 +444,30 @@ func (s *IntegrationTestSuite) TestSuperfluidVoting() {
...
+	// Need to json unparshal the params because they are stored including quotes
+	paramA := nodeA.QueryParams(ibcratelimittypes.ModuleName, string(ibcratelimittypes.KeyContractAddress))
+	json.Unmarshal([]byte(paramA), &paramA)
+	paramB := nodeB.QueryParams(ibcratelimittypes.ModuleName, string(ibcratelimittypes.KeyContractAddress))
+	json.Unmarshal([]byte(paramB), &paramB)
+
+	// When upgrading to v15, we want to make sure that the rate limits have been set.
+	quotas, err := nodeA.QueryWasmSmartArray(paramA, `{"get_quotas": {"channel_id": "any", "denom": "ibc/E6931F78057F7CC5DA0FD6CEF82FF39373A6E0452BF1FD76910B93292CF356C1"}}`)
+	s.Require().Len(quotas, 2)
+	s.Require().NoError(err)
```
Comment (p0mvn) [inline tests/e2e/e2e_test.go:470]: Should we repeat this check for both chains A and B?

The reason is that for chain A we overwrite the quotas in `TestIBCTokenTransferRateLimiting` so making this check against `nodeA` only does not validate the correctness of the upgrade

**Review by p0mvn — APPROVED** [review]
LGTM
